File: py/2021/day2.py
def read_input(year: int, day: int):
    with open(f"input/{year}/{day}", "r") as f:
        return f.read()


inputs = [x for x in read_input(2021, 2).split("\n")]


def part_one():
    depth = 0
    horizontal = 0

    for line in inputs:
        (command, value) = line.split(" ")
        if command == "forward":
            horizontal += int(value)
        elif command == "up":
            depth -= int(value)
        else:
            depth += int(value)

    print(f"Part One: {depth * horizontal}")


def part_two():
    aim = 0
    depth = 0
    horizontal = 0

    for line in inputs:
        (command, value) = line.split(" ")
        if command == "forward":
            horizontal += int(value)
            depth += int(value) * aim
        elif command == "up":
            aim -= int(value)
        else:
            aim += int(value)

    print(f"Part Two: {depth * horizontal}")


part_one()
part_two()

File: cpp/src/2021/day2.cpp
#include "advent-of-code.h"

void part_one(std::vector<std::string> &raw_input);
void part_two(std::vector<std::string> &raw_input);

int main() {
  std::vector<std::string> raw_input{read_input(2021, 2)};

  part_one(raw_input);
  part_two(raw_input);

  return 0;
}

void part_one(std::vector<std::string> &raw_input) {
  int depth{0};
  int horizontal{0};

  for (auto line : raw_input) {
    auto values{split(line, ' ')};
    std::string command{values[0]};
    int value{std::stoi(values[1])};

    if (command == "forward") {
      horizontal += value;
    } else if (command == "up") {
      depth -= value;
    } else {
      depth += value;
    }
  }

  std::cout << "Part One: " << depth * horizontal << std::endl;
}

void part_two(std::vector<std::string> &raw_input) {
  int aim{0};
  int depth{0};
  int horizontal{0};

  for (auto line : raw_input) {
    auto values{split(line, ' ')};
    std::string command{values[0]};
    int value{std::stoi(values[1])};

    if (command == "forward") {
      horizontal += value;
      depth += value * aim;
    } else if (command == "up") {
      aim -= value;
    } else {
      aim += value;
    }
  }

  std::cout << "Part Two: " << depth * horizontal << std::endl;
}
